plugins/storage/pyparts/ceph_cluster_checks.py

```diff
@@ -336,6 +336,76 @@ def get_crushmap_mixed_buckets(self):
             issue = issue_types.CephCrushWarning(msg)
             issue_utils.add_issue(issue)
 
+    def _is_bucket_imbalanced(self, buckets, start_bucket_id, failure_domain):
+        """
+        Recursively determine if a tree is balanced at the given failure domain
+        """
+        unbalanced = False
+        weight = -1
+
+        for item in buckets[start_bucket_id]["items"]:
+            if buckets[item["id"]]["type_name"] != failure_domain:
+                unbalanced = self._is_bucket_imbalanced(buckets,
+                                                        item["id"],
+                                                        failure_domain)
+                if unbalanced:
+                    return unbalanced
+            else:
+                if weight == -1:
+                    weight = item["weight"]
+                else:
+                    if weight != item["weight"]:
+                        unbalanced = True
+                        return unbalanced
+
+        return unbalanced
+
+    def check_crushmap_equal_buckets(self):
+        """
+        Report when buckets of the failure domain type in a
+        CRUSH rule referenced tree are unbalanced.
+
+        Uses the trees and failure domains referenced in the
+        CRUSH rules, and checks that all buckets of the failure
+        domain type in the referenced tree are equal.
+        """
+        osd_crush_dump = self.cli.ceph_osd_crush_dump_json_decoded()
+        if not osd_crush_dump:
+            return
+
+        buckets = {}
+        for bucket in osd_crush_dump["buckets"]:
+            bid = bucket["id"]
+            buckets[bid] = bucket
```

Comment:
can replace the above with a one-liner e.g.

buckets = {b['id']: b for b in osd_crush_dump['buckets']}

Reply:
also since you are not actually using buckets until after processing the rules you could move it to line 394

Reply:
Moved it to 390, don't want to rebuild the buckets each time through the loop.